torch/distributed/_symmetric_memory/_autograd.py

```diff
@@ -134,3 +134,145 @@ def backward(  # type: ignore[no-untyped-def]
             group_name=ctx.group_name,
         )
         return grad_input, None, None, None
+
+
+class AllToAllVDev2dOffset(torch.autograd.Function):
+    """
+    Autograd function for `all_to_all_vdev_2d_offset`
+    Usage:
+        >>> # xdoctest: +SKIP("undefined vars")
+        >>> # Initialize
+        >>> AllToAllVDev2dOffset.init(max_output_len, alignment)
+        >>> # Forward
+        >>> output, out_splits_offsets = AllToAllVDev2dOffset.apply(
+        ...     input, in_splits_offsets, group_name
+        ... )
+        >>> # Backward
+        >>> s = output.sum()
+        >>> s.backward()
+        >>> print(input.grad)
+    """
+
+    # Maximum output length (need to be set before use of AllToAllVDev2dOffset)
+    max_output_len: Optional[int] = None
+    # Alignment of the input offsets
+    alignment: Optional[int] = None
+    # A symmetric memory holding the grad_output during backward
+    grad_output_buf: Optional[torch.Tensor] = None
+
+    @staticmethod
+    def init(
+        max_output_len: int,
+        alignment: int,
+    ) -> None:
+        """
+        Init function for `AllToAllVDev2dOffset`. Must be called before use.
+        Args:
+            `max_output_len`: maximum output length, must be larger than the actual output length.
+        """
+        AllToAllVDev2dOffset.max_output_len = max_output_len
+        AllToAllVDev2dOffset.alignment = alignment
+
+    @staticmethod
+    def forward(  # type: ignore[no-untyped-def]
+        ctx,
+        input: torch.Tensor,
+        in_splits_offsets: torch.Tensor,
+        group_name: str,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """
+        Functionality is the same as `all_to_all_vdev_2d_offset` but with `output` and
+        `out_splits_offsets` returned instead of modified in place.
+        """
+        if AllToAllVDev2dOffset.max_output_len is None:
+            raise RuntimeError(
+                "Please set max output length via `AllToAllVDev2dOffset.init(max_output_len)`"
+            )
+
+        # Allocate output buffer
+        output = symm_mem.empty(
```

Comment:
It would be great if you could reuse these buffers for subsequent fowards, they can get pretty large right?